### PR TITLE
`_restricted_classmethod`: add wrapper, to allow inspection

### LINKED ISSUE #19329: Issue with _restricted_classmethod_impl Triggering TypeError on inspect.getmembers Call
### Bug description

When using `inspect.getmembers` on a `LightningModule` instance (or any class using `_restricted_classmethod_impl`), a `TypeError` is incorrectly raised. This issue was introduced in PR #18432 ("Raise exception when load_from_checkpoint called from instance") and is present in version 2.1.1 onwards. The current implementation triggers the check during attribute access rather than at the time of method call.

### What version are you seeing the problem on?

v2.1, master

### How to reproduce the bug

```python
import inspect
from pytorch_lightning.utilities.model_helpers import _restricted_classmethod_impl

class ExampleClass:
    @_restricted_classmethod_impl
    def example_method(cls):
        print("hello")

if __name__ = "__main__":
    inspect.getmembers(ExampleClass())
```

### Error messages and logs

```
Traceback (most recent call last):
  File "PATH/file.py", line 10, in <module>
    inspect.getmembers(ExampleClass())
  File "/home/USER/.conda/envs/python311/lib/python3.11/inspect.py", line 595, in getmembers
    return _getmembers(object, predicate, getattr)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/USER/.conda/envs/python311/lib/python3.11/inspect.py", line 573, in _getmembers
    value = getter(object, key)
            ^^^^^^^^^^^^^^^^^^^
  File "/home/USER/.conda/envs/python311/lib/python3.11/site-packages/pytorch_lightning/utilities/model_helpers.py", line 93, in __get__
    raise TypeError(
TypeError: The classmethod `ExampleClass.example_method` cannot be called on an instance. Please call it on the class type and make sure the return value is used.
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow): LightningModule
#- PyTorch Lightning Version (e.g., 1.5.0): 2.1.3
#- Lightning App Version (e.g., 0.5.2): 
#- PyTorch Version (e.g., 2.0): 2.1.2
#- Python version (e.g., 3.9): 3.11
#- OS (e.g., Linux): Linux
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source): pip
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

**Suggested Solution**

Modify `_restricted_classmethod_impl` to perform the instance check only when the method is actually called. Proposed change:

```python
class _restricted_classmethod_impl(Generic[_T, _P, _R_co]):
    # ... [existing class documentation]

    def __init__(self, method: Callable[Concatenate[_T, _P], _R_co]) -> None:
        self.method = method

    def __get__(self, instance: Optional[_T], cls: Type[_T]) -> Callable[_P, _R_co]:
        def wrapper(*args, **kwargs):
            # Workaround for https://github.com/pytorch/pytorch/issues/67146
            is_scripting = any(os.path.join("torch", "jit") in frameinfo.filename for frameinfo in inspect.stack())
            if instance is not None and not is_scripting:
                raise TypeError(
                    f"The classmethod `{cls.__name__}.{self.method.__name__}` cannot be called on an instance."
                    " Please call it on the class type and make sure the return value is used."
                )
            return self.method(cls, *args, **kwargs)
        
        return wrapper

### PR DESCRIPTION
**Pull request by lauritsf**
## What does this PR do?

This PR resolves an issue where using `inspect.getmembers` on a `LightningModule` instance (or any class using `_restricted_classmethod_impl`) would incorrectly raise a `TypeError`. The modification introduces a wrapper in the `_restricted_classmethod_impl` to ensure that the method can be inspected but not called on an instance. This change maintains the intended functionality while improving compatibility with introspection tools  like `inspect.getmembers`.

Fixes #19329 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19332.org.readthedocs.build/en/19332/

### CODE REVIEW

File: tests/tests_pytorch/utilities/test_model_helpers.py [hunk 1]
```diff
@@ -71,6 +71,11 @@ def test_restricted_classmethod():
 
     RestrictedClass.restricted_cmethod()  # no exception
 
+    restricted_method = RestrictedClass().restricted_cmethod  # no exception
+
+    with pytest.raises(TypeError, match="cannot be called on an instance"):
+        restricted_method()
```
Comment (awaelchli) [inline tests/tests_pytorch/utilities/test_model_helpers.py:77]: These lines can directly replace the lines above. Might also want to throw in a `inspect.getmembers(RestrictedClass)  # no exception`

**Review by awaelchli — APPROVED** [review]
Looks great. Thanks for the quick fix. Would you mind adding a line in `src/lightning/pytorch/CHANGELOG.md` as well?

File: tests/tests_pytorch/utilities/test_model_helpers.py (hunk 1 above)
Comment (awaelchli) [reply tests/tests_pytorch/utilities/test_model_helpers.py]: Perfect!

File: src/lightning/pytorch/utilities/model_helpers.py [hunk 2]
```diff
@@ -108,18 +107,22 @@ class _restricted_classmethod_impl(Generic[_T, _P, _R_co]):
...
-        # Workaround for https://github.com/pytorch/pytorch/issues/67146
-        is_scripting = any(os.path.join("torch", "jit") in frameinfo.filename for frameinfo in inspect.stack())
-        if instance is not None and not is_scripting:
-            raise TypeError(
-                f"The classmethod `{cls.__name__}.{self.method.__name__}` cannot be called on an instance."
-                " Please call it on the class type and make sure the return value is used."
-            )
-        return MethodType(self.method, cls)
+        # The wrapper ensures that the method can be inspected, but not called on an instance
+        def wrapper(*args: Any, **kwargs: Any) -> _R_co:
```
Comment (awaelchli) [inline src/lightning/pytorch/utilities/model_helpers.py:115]: The docs build complains, I think we need to use `@functools.wraps(self.method)` here

**Review by Borda — APPROVED** [review]

**Pull request merged**